src/Native/System.Native/pal_networking.cpp

```diff
@@ -1244,6 +1253,42 @@ extern "C" Error SetIPv6MulticastOption(int32_t socket, int32_t multicastOption,
     return err == 0 ? PAL_SUCCESS : ConvertErrorPlatformToPal(errno);
 }
 
+#if defined(__APPLE__) && __APPLE__
+static int32_t GetMaxLingerTime()
+{
+    static volatile int32_t MaxLingerTime = -1;
+
+    // OS X does not define the linger time in seconds by default, but in ticks.
+    // Furthermore, when SO_LINGER_SEC is used, the value is simply scaled by
+    // the number of ticks per second and then the result is used to set the
+    // underlying linger time. Unfortunately, the underlying linger time is
+    // stored as a `short` and out-of-range values are simply truncated to fit
+    // within 16 bits and then reinterpreted as 2's complement signed integers.
+    // This results in some *very* strange behavior and a rather low limit for
+    // the linger time. Instead of admitting this behavior, we determine the
+    // maximum linger time in seconds and return an error if the input is out
+    // of range.
+    int32_t maxLingerTime = MaxLingerTime;
+    if (maxLingerTime == -1)
+    {
+        long ticksPerSecond = sysconf(_SC_CLK_TCK);
+        maxLingerTime = static_cast<int32_t>(32767 / ticksPerSecond);
```

Comment:
Let me check. We can't check `sizeof(linger::l_linger)` since it's still an `int`. There's something in a different header file that I used to root cause the bug; let me see if that structure is actually available at compile-time (i.e. it's not #ifdef'd out).